src/freenet/l10n/BaseL10n.java

```diff
@@ -591,14 +592,69 @@ public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, St
 	 * @param values For each entry in the previous array parameter, this array specifies the {@link HTMLNode} with which it shall be replaced. 
 	 */
 	public void addL10nSubstitution(HTMLNode node, String key, String[] patterns, HTMLNode[] values) {
-		String value = getString(key);
-		addL10nSubstitutionInner(node, key, value, patterns, values);
+        List<HTMLNode> newContent = getL10nSubstitution(key, patterns, values);
+        if (newContent != null) {
+            node.addChildren(newContent);
+        }
 	}
+    
+    /**
+     * Return whether a l10n string possibly containing substitution variables can be parsed.
+     * Intended for use in tests.
+     */
+    boolean isL10nValid(String key, String value) {
+        String[] patterns = new String[0];
+        HTMLNode[] values = new HTMLNode[0];
+        List<HTMLNode> content = tryGetL10nSubstitution(key, value, patterns, values);
+        return content != null;
+    }
+    
+    /**
+     * Looks up a l10n string and performs substitutions to generate a list of {@link HTMLNode}s.
+     */
+    private List<HTMLNode> getL10nSubstitution(String key, String[] patterns, HTMLNode[] values) {
+        List<HTMLNode> newContent = null;
+        // first try the value in the current language
+        String value = getString(key, true);
+        if (value != null) {
+            newContent = tryGetL10nSubstitution(key, value, patterns, values);
+        }
+        // then try the fallback
+        if (newContent == null) {
+            String fallbackValue = getDefaultString(key);
+            if (!fallbackValue.equals(value)) {
+                newContent = tryGetL10nSubstitution(key, fallbackValue, patterns, values);
+            }
+        }
+        // if that doesn't work, just use the key itself
+        if (newContent == null) {
+            newContent = Collections.singletonList(new HTMLNode("#", key));
+        }
+        return newContent;
+    }
+    
+    /**
+     * Performs substitutions on a l10n string to generate a list of {@link HTMLNode}s. Returns
+     * null if an error is encountered.
+     */
+    private List<HTMLNode> tryGetL10nSubstitution(String key, String value,
+            String[] patterns, HTMLNode[] values) {
```

Comment:
Looks like this method only performs substitution. There's no need to pass the `key` here.